### PR TITLE
fix: curp wal

### PR DESCRIPTION
**Pull request by bsbds**
Depends-On: #687 

Some fixes and improvements for the WAL storage

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: crates/curp/src/server/storage/wal/pipeline.rs [hunk 1]
```diff
@@ -34,6 +34,9 @@ pub(super) struct FilePipeline {
     stopped: Arc<AtomicBool>,
     /// Join handle of the allocation task
     file_alloc_task_handle: Option<JoinHandle<()>>,
+    #[cfg_attr(not(madsim), allow(unused))]
```
Comment (Phoenix500526) [inline crates/curp/src/server/storage/wal/pipeline.rs:37]: Please use `#[cfg(madsim)]` to replace it.

File: crates/curp/src/server/storage/wal/pipeline.rs [hunk 2]
```diff
@@ -80,41 +83,19 @@ impl FilePipeline {
...
-            });
-
             Self {
                 dir,
                 file_size,
-                file_iter: Some(file_rx.into_iter()),
+                file_iter: None,
                 stopped,
                 file_alloc_task_handle: None,
+                file_count: 0,
```
Comment (Phoenix500526) [inline crates/curp/src/server/storage/wal/pipeline.rs:98]: Please add `#[cfg(madsim)]` for the `file_count` field.

File: crates/curp/src/server/storage/wal/codec.rs [hunk 3]
```diff
@@ -282,17 +285,18 @@ where
                     .unwrap_or_else(|_| unreachable!("serialization should never fail"));
                 let len = entry_bytes.len();
                 assert_eq!(len >> 56, 0, "log entry length: {len} too large");
-                let len_bytes = len.to_le_bytes().into_iter().take(7);
-                let header = std::iter::once(self.frame_type()).chain(len_bytes);
-                header.chain(entry_bytes).collect()
+                let mut bytes = Vec::with_capacity(1 + 7 + entry_bytes.len());
```
Comment (Phoenix500526) [inline crates/curp/src/server/storage/wal/codec.rs:288]: Can we replace `1` with `std::mem::size_of_val(&self.frame_type())`?  And using a constant to replace the magic number `7`.

File: crates/curp/src/server/storage/wal/codec.rs [hunk 4]
```diff
@@ -282,17 +285,18 @@ where
...
             }
             DataFrame::SealIndex(index) => {
                 assert_eq!(index >> 56, 0, "log index: {index} too large");
-                // use the first 7 bytes
-                let index_bytes = index.to_le_bytes().into_iter().take(7);
-                std::iter::once(self.frame_type())
-                    .chain(index_bytes)
-                    .collect()
+                let mut bytes = index.to_le_bytes();
+                bytes.rotate_right(1);
```
Comment (Phoenix500526) [inline crates/curp/src/server/storage/wal/codec.rs:297]: Same as above

Comment (bsbds) [reply crates/curp/src/server/storage/wal/codec.rs:315]: I think it's clear enough to use number here.

File: crates/curp/src/server/storage/wal/mod.rs [hunk 5]
```diff
@@ -157,10 +157,9 @@ where
     /// Send frames with fsync
     #[allow(clippy::pattern_type_mismatch)] // Cannot satisfy both clippy
     pub(super) fn send_sync(&mut self, item: Vec<DataFrame<'_, C>>) -> io::Result<()> {
-        let last_segment = self
```
[the remark this answers is not in the document]
Comment (bsbds) [reply crates/curp/src/server/storage/wal/mod.rs]: it's previous used in debugging, removed

File: crates/curp/src/server/storage/wal/pipeline.rs (hunk 1 above)
Comment (bsbds) [reply crates/curp/src/server/storage/wal/pipeline.rs]: done

**Review by Phoenix500526 — APPROVED** [review]
LGTM, please squash these commits into one.

**Comment by CrystalAnalyst** [comment]
LGTM

**Review by CrystalAnalyst — APPROVED** [review]

**Review by Phoenix500526 — CHANGES_REQUESTED** [review]
Please squash these commits into one

**Comment by Phoenix500526** [comment]
@Mergifyio rebase

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Pull request merged**